In [151]:
import plotly.express as px
import plotly as pl
import pandas as pd 
import json 
import numpy as np 
from bs4 import BeautifulSoup

In [152]:
with open("user_data_tiktok.json") as j:
    tiktok = json.load(j)

### The section below is exploration. The next markdown outlines the work for Milestone 1 

In [153]:
tiktok

{'Comment': {'Comments': {'App': 1,
   'CommentsList': [{'date': '2026-03-15 15:44:38',
     'comment': 'Disneyland?',
     'photo': 'N/A',
     'video': 'N/A',
     'url': ''},
    {'date': '2026-03-15 13:23:43',
     'comment': 'My fave is chip and dale!',
     'photo': 'N/A',
     'video': 'N/A',
     'url': ''},
    {'date': '2026-03-13 02:42:13',
     'comment': 'While I am sorry you’ve had to deal with all that, this gave me the greatest laugh',
     'photo': 'N/A',
     'video': 'N/A',
     'url': ''},
    {'date': '2026-03-12 20:49:59',
     'comment': 'Love this but hate the idea of you doing child labor 😂',
     'photo': 'N/A',
     'video': 'N/A',
     'url': ''},
    {'date': '2026-03-10 02:05:01',
     'comment': 'My favorite Disney characters are chip and dale bc your can’t separate the icons 🥰 @Gabe',
     'photo': 'N/A',
     'video': 'N/A',
     'url': ''},
    {'date': '2026-02-14 18:18:37',
     'comment': 'Iconic',
     'photo': 'N/A',
     'video': 'N/A',
     'url

In [154]:
type(tiktok)

dict

In [155]:
tiktok.keys()

dict_keys(['Comment', 'Direct Message', 'Income+ Wallet', 'Likes and Favorites', 'Location Review', 'Post', 'Profile And Settings', 'TikTok Live', 'TikTok Shop', 'Your Activity'])

In [156]:
tiktok["Your Activity"]

{'Ad Interests': {'AdInterestCategories': ','},
 'Ads Visit History': {'AdsVisitHistoryList': [{'CreateTime': '2025-11-08 15:32:14',
    'AdTitle': 'Christmas : Target',
    'AdLink': 'https://www.target.com/c/christmas/-/N-5xt30?TCID=SMP-1847411527626769-1847412809923602-1847962512477265&ttclid=E_C_P_CtwBFdvlGnauO1CQVvQ11QCOfZ4OS4gj08NnDxw73Q7bqyrtURS8eYChII8bZU0Kb6tiYuXPFZhv96pOSIl661vsqqBMjGoRDO3FFqicxp4IdJVqucLF6R-T1DU0wnBhpy-qViXXO9anATKYd_Qoc0tTx47Ou01_WWe_n5zoKrR9BWX025XVYfXExtbn-fZxbIx6QsjNBWYTC8GIoWl9JKAGYomAus_Wbm5nA2R8nekJJ9cgISqn1OktsRy4xsXBDPNTG2p-qZcBR4t1i0yhTVTJCQTMncDPIcNUeIAvXxIEdjIuMA'},
   {'CreateTime': '2025-11-11 02:02:01',
    'AdTitle': "A.1. Sauce: It's Not Meatloaf, It's Beefcake Campaign | United States | Kraft Heinz",
    'AdLink': 'https://www.kraftheinz.com/a1beefcakes?utm_medium=paidsocial&utm_source=tiktok&utm_campaign=1847801379103778&utm_adgroup=1847800631884978&utm_content=1847954492354609&ttclid=E_C_P_Ct0BzBm9idU_zLgMR6zpuJPjrp4YmEK0mhN-wvf8uN0Q6kBfx

In [157]:
len(tiktok["Your Activity"])

15

In [158]:
df_tiktok = pd.DataFrame(tiktok["Your Activity"])
df_tiktok.head()

,Ad Interests,Ads Visit History,Donation,Fundraiser,Hashtag,Instant Form Ads Responses,Login History,Off TikTok Activity,Purchases,Reposts,Searches,Share History,Status,Stickers,Watch History
AdInterestCategories,",",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
AdsVisitHistoryList,NaN,"[{'CreateTime': '2025-11-08 15:32:14', 'AdTitl...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
DonationList,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
FundraiserList,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
HashtagList,NaN,NaN,NaN,NaN,"[{'HashtagName': 'rupaul', 'HashtagLink': 'htt...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [159]:
len(df_tiktok)

16

In [160]:
df_tiktok

,Ad Interests,Ads Visit History,Donation,Fundraiser,Hashtag,Instant Form Ads Responses,Login History,Off TikTok Activity,Purchases,Reposts,Searches,Share History,Status,Stickers,Watch History
AdInterestCategories,",",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
AdsVisitHistoryList,NaN,"[{'CreateTime': '2025-11-08 15:32:14', 'AdTitl...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
DonationList,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
FundraiserList,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
HashtagList,NaN,NaN,NaN,NaN,"[{'HashtagName': 'rupaul', 'HashtagLink': 'htt...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
ResponsesList,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
LoginHistoryList,NaN,NaN,NaN,NaN,NaN,NaN,"[{'Date': '2024-12-26 00:01:54', 'IP': '75.38....",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
OffTikTokActivityDataList,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"[{'TimeStamp': '2025-11-28 19:30:19', 'Source'...",NaN,NaN,NaN,NaN,NaN,NaN,NaN
SendGifts,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,{'SendGifts': None},NaN,NaN,NaN,NaN,NaN,NaN
BuyGifts,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,{'BuyGifts': None},NaN,NaN,NaN,NaN,NaN,NaN


In [161]:
df_tiktok['Searches'].iloc[11]

[{'Date': '2026-03-20 00:41:44', 'SearchTerm': 'Papa Giannopoulos'},
 {'Date': '2026-03-18 14:00:08', 'SearchTerm': 'City pop'},
 {'Date': '2026-03-18 02:02:02',
  'SearchTerm': 'lemon chicken cheesecake factory'},
 {'Date': '2026-03-18 00:20:52', 'SearchTerm': 'Shawn Dalton'},
 {'Date': '2026-03-16 03:36:43', 'SearchTerm': 'Kyle Bee'},
 {'Date': '2026-03-16 03:25:19', 'SearchTerm': 'olliebollen'},
 {'Date': '2026-03-16 01:29:14', 'SearchTerm': 'azazie becka dress'},
 {'Date': '2026-03-16 00:40:49', 'SearchTerm': 'gavin newsom vs barack obama'},
 {'Date': '2026-03-15 21:36:58', 'SearchTerm': 'Cinnamon azazie color'},
 {'Date': '2026-03-15 21:36:52', 'SearchTerm': 'Cinnamon paprika color'},
 {'Date': '2026-03-15 21:35:31', 'SearchTerm': 'azazie paprika color'},
 {'Date': '2026-03-15 15:44:17', 'SearchTerm': 'city'},
 {'Date': '2026-03-15 15:44:01', 'SearchTerm': 'My city'},
 {'Date': '2026-03-15 04:15:27', 'SearchTerm': 'Leaked labs'},
 {'Date': '2026-03-15 04:14:23', 'SearchTerm': 'the

In [162]:
len(df_tiktok['Searches'].iloc[11])

244

In [163]:
### df_tiktok['Share History'].iloc[12]

In [164]:
len(df_tiktok['Share History'].iloc[12])

3812

In [165]:
len(df_tiktok['Watch History'].iloc[15])

78194

In [166]:
### df_tiktok['Watch History'].iloc[15]

In [167]:
len(df_tiktok['Ad Interests'].iloc[0])

1

In [168]:
df_tiktok['Ad Interests'].iloc[0]

','

# Milestone 1 
So... I thought tidy data by milestone 1 was possible. BOY IT WAS NOT with my current skill set. So here's the NEW milestone: TIKTOK data is tidy. Webscraping will be crucial for the Instagram portion and... I'm going to wait until the lecture to make that happen. 
I've also refined my questions I'm hoping to answer within understanding how students use tiktok vs how they use Instagram.  
### First, what advertising terms/interests do each platform assign to them? 
What are the big terms within each platform? 
     Somehow this is NOT in your activity on Tiktok's downloadable data. "Ad interest" is in the data set but it's a COMMA? They just let you see it without having to download which is nice but it's NOT in the nice neat json file. So for this one I will need to add it the Tiktok Tidy Data Frame. 
### Second, Watch times, how do they differ between platforms? 
I'm HOPING this isn't a pain in the butt. 
Notes from convo with Dr. Silva: Date Time delta, PD.date 
### Third, Activity times 
 When are students most likely to be on which platform? 
Dates are nice but what would be the most helpful would be what day of the week is the most active on each platform? Bonus points for time of day. 
Notes from convo with Dr. Silva: New column, full date time, hour element 24 hour portion, aggrigate, day element, day element, Heat map, Look at Timestamp documentation for pandas, you can also Grab weekday vs weekend elements. Concat take all of these rows and add them. Git hub parsing, Large lanuage models to look for tools 

In [169]:
with open("IG recommended_topics.html", encoding= "utf-8") as IGTopics: 
    ig_dat1 = BeautifulSoup(IGTopics.read(), "html.parser")

In [170]:
ig_dat1

<html><head><meta content="text/html; charset=utf-8" http-equiv="Content-Type"/><base href="../../"/><style nonce="1lSzB6Zr" type="text/css">html{touch-action:manipulation}body{background:#fff;color:#1c1e21;direction:ltr;line-height:1.34;margin:0;padding:0;unicode-bidi:embed}body,button,input,label,select,td,textarea{font-family:Helvetica,Arial,sans-serif;font-size:12px}h1,h2,h3,h4,h5,h6{color:#1c1e21;font-size:13px;font-weight:600;margin:0;padding:0}h1{font-size:14px}h4,h5,h6{font-size:12px}p{margin:1em 0}b,strong{font-weight:600}a{color:#385898;cursor:pointer;text-decoration:none}button{margin:0}a:hover{text-decoration:underline}img{border:0px}td,td.label{text-align:left}dd{color:#000}dt{color:#606770}ul{list-style-type:none;margin:0;padding:0}abbr{border-bottom:none;text-decoration:none}hr{background:#dadde1;border-width:0;color:#dadde1;height:1px}
form{margin:0;padding:0}label{color:#606770;cursor:default;font-weight:600;vertical-align:middle}label input{font-weight:400}textarea,.i

Well.. I'm not going to touch that just yet

In [171]:
tablett2= pd.DataFrame(df_tiktok['Watch History'].iloc[15])

In [172]:
tablett2.head()

,Date,Link
0,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...
1,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...
2,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...
3,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...
4,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...


Okay here are the columns I need: 
1. Platform ID (in this case it will always be Tiktok)
2. Person ID (in this case it will be A)
3. Date of the video watched (OKAY WE'RE HALF WAY THERE)
4. What day of the week correlates with that date 
5. Time of the video watched (for 0 it would be 12:33:12)
6. Time of each video watched 

In [173]:
tablett2.insert(0, "Platform", ["Tiktok"], True)
print(tablett2)

ValueError: Length of values (1) does not match length of index (78194)

Okay, troubleshooting here. Have to make "Tiktok" appear 78194 times!!!! 

In [ ]:
tablett2.assign(Platform="Tiktok")

,Date,Link,Platform
0,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,Tiktok
1,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,Tiktok
2,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,Tiktok
3,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,Tiktok
4,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,Tiktok
...,...,...,...
78189,2025-09-22 00:40:31,https://www.tiktokv.com/share/video/7540322581...,Tiktok
78190,2025-09-22 00:39:59,https://www.tiktokv.com/share/video/7550818580...,Tiktok
78191,2025-09-22 00:39:53,https://www.tiktokv.com/share/video/7552242622...,Tiktok
78192,2025-09-22 00:39:44,https://www.tiktokv.com/share/video/7552286495...,Tiktok


GOD BLESS THE INTERNET! Here's where it gave me the code to do that https://stackoverflow.com/questions/44657226/how-to-insert-a-new-column-with-repeated-values-into-a-pandas-table omg so much easier

In [ ]:
tablett2.assign(Platform='Tiktok')[['Platform'] + tablett2.columns.tolist()]

,Platform,Date,Link
0,Tiktok,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...
1,Tiktok,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...
2,Tiktok,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...
3,Tiktok,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...
4,Tiktok,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...
...,...,...,...
78189,Tiktok,2025-09-22 00:40:31,https://www.tiktokv.com/share/video/7540322581...
78190,Tiktok,2025-09-22 00:39:59,https://www.tiktokv.com/share/video/7550818580...
78191,Tiktok,2025-09-22 00:39:53,https://www.tiktokv.com/share/video/7552242622...
78192,Tiktok,2025-09-22 00:39:44,https://www.tiktokv.com/share/video/7552286495...


In [ ]:
tablett2.assign(Platform="Tiktok", Person="A")[['Platform', 'Person'] + tablett2.columns.tolist()]

,Platform,Person,Date,Link
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...
...,...,...,...,...
78189,Tiktok,A,2025-09-22 00:40:31,https://www.tiktokv.com/share/video/7540322581...
78190,Tiktok,A,2025-09-22 00:39:59,https://www.tiktokv.com/share/video/7550818580...
78191,Tiktok,A,2025-09-22 00:39:53,https://www.tiktokv.com/share/video/7552242622...
78192,Tiktok,A,2025-09-22 00:39:44,https://www.tiktokv.com/share/video/7552286495...


Full disclosure, I tried a bunch of different iterations for the above. I had the above pieces but the wrong syntax so Claude helped bc I had () where I should have had commas.

In [ ]:
Tempsave= tablett2.assign(Platform="Tiktok", Person="A")[['Platform', 'Person'] + tablett2.columns.tolist()]

...So I know if I keep making data frames this thing will be a PAIN to replicate for multiple platforms for multiple people, BUT while i'm still messing around this is my temp plan. So hopefully once I finish it I can streamline it into a process that actually can be more replicatable 

Now I need to break up the date from the time

In [ ]:
type(Tempsave["Date"].iloc[0])

str

In [ ]:
Tempsave["datetime"] = pd.to_datetime(Tempsave["Date"])
Tempsave.head()

,Platform,Person,Date,Link,datetime
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,2026-03-20 12:33:12
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,2026-03-20 12:32:59
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,2026-03-20 12:02:37
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,2026-03-20 12:02:36
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,2026-03-20 11:53:03


Great, converted my string to date time but it got added to the end. I'm going to reorder the columns when everything is done bc that's cosmetic at this point. 

In [ ]:
Tempsave["date"]=Tempsave["datetime"].dt.date
Tempsave.head()

,Platform,Person,Date,Link,datetime,date
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,2026-03-20 12:33:12,2026-03-20
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,2026-03-20 12:32:59,2026-03-20
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,2026-03-20 12:02:37,2026-03-20
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,2026-03-20 12:02:36,2026-03-20
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,2026-03-20 11:53:03,2026-03-20


Okay... now I have two columns called date... The old one called "Date" which is a string, and the new one called "date"which is in date time. This will be annoying but maybe after I'm done I can figure out how to clean that one up lol. Maybe I can rename "Date" to like "string" or something earlier on??? Since I need date time format for graphing?

In [ ]:
Tempsave["Time"]=Tempsave["datetime"].dt.time
Tempsave

,Platform,Person,Date,Link,datetime,date,Time
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,2026-03-20 12:33:12,2026-03-20,12:33:12
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,2026-03-20 12:32:59,2026-03-20,12:32:59
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,2026-03-20 12:02:37,2026-03-20,12:02:37
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,2026-03-20 12:02:36,2026-03-20,12:02:36
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,2026-03-20 11:53:03,2026-03-20,11:53:03
...,...,...,...,...,...,...,...
78189,Tiktok,A,2025-09-22 00:40:31,https://www.tiktokv.com/share/video/7540322581...,2025-09-22 00:40:31,2025-09-22,00:40:31
78190,Tiktok,A,2025-09-22 00:39:59,https://www.tiktokv.com/share/video/7550818580...,2025-09-22 00:39:59,2025-09-22,00:39:59
78191,Tiktok,A,2025-09-22 00:39:53,https://www.tiktokv.com/share/video/7552242622...,2025-09-22 00:39:53,2025-09-22,00:39:53
78192,Tiktok,A,2025-09-22 00:39:44,https://www.tiktokv.com/share/video/7552286495...,2025-09-22 00:39:44,2025-09-22,00:39:44


Now to turn "date" with the lowercase D into a day

In [ ]:
Tempsave["Day"]=Tempsave["datetime"].dt.day
Tempsave.head()

,Platform,Person,Date,Link,datetime,date,Time,Day
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,2026-03-20 12:33:12,2026-03-20,12:33:12,20
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,2026-03-20 12:32:59,2026-03-20,12:32:59,20
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,2026-03-20 12:02:37,2026-03-20,12:02:37,20
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,2026-03-20 12:02:36,2026-03-20,12:02:36,20
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,2026-03-20 11:53:03,2026-03-20,11:53:03,20


Well just changing it to day was worth a shot lol

In [ ]:
Tempsave["Day"]=Tempsave["datetime"].dt.day_name
Tempsave.head()

,Platform,Person,Date,Link,datetime,date,Time,Day
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,2026-03-20 12:33:12,2026-03-20,12:33:12,<bound method PandasDelegate._add_delegate_acc...
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,2026-03-20 12:32:59,2026-03-20,12:32:59,<bound method PandasDelegate._add_delegate_acc...
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,2026-03-20 12:02:37,2026-03-20,12:02:37,<bound method PandasDelegate._add_delegate_acc...
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,2026-03-20 12:02:36,2026-03-20,12:02:36,<bound method PandasDelegate._add_delegate_acc...
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,2026-03-20 11:53:03,2026-03-20,11:53:03,<bound method PandasDelegate._add_delegate_acc...


In [ ]:
Tempsave["Day"].iloc[0]

<bound method PandasDelegate._add_delegate_accessors.<locals>._create_delegator_method.<locals>.f of <pandas.core.indexes.accessors.DatetimeProperties object at 0x337930d70>>

In [ ]:
Tempsave["Day"]=Tempsave["datetime"].dt.day_name()
Tempsave.head()

,Platform,Person,Date,Link,datetime,date,Time,Day
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,2026-03-20 12:33:12,2026-03-20,12:33:12,Friday
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,2026-03-20 12:32:59,2026-03-20,12:32:59,Friday
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,2026-03-20 12:02:37,2026-03-20,12:02:37,Friday
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,2026-03-20 12:02:36,2026-03-20,12:02:36,Friday
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,2026-03-20 11:53:03,2026-03-20,11:53:03,Friday


Oh pandas... the () make all the difference. (had to google that one)
Well... just glad it worked. Now I need to do the delta for time.... and hopefully that's easy

In [ ]:
Tempsave["Watch Time"] = pd.to_timedelta(Tempsave["Time"])

TypeError: Invalid type for timedelta scalar: <class 'datetime.time'>

OH MY GOD YOU CAN'T DETLA TIME A TIMECODE?????? IT HAS TO BE A STRING????? I CONVERTED IT TO DATE TIME FROM A STRING. WE'VE BEEN TO THE MOON AND WE CAN'T DO THAT?
![Image](https://merriam-webster.com/assets/mw/images/article/art-wap-article-main/headdesk-2560-44956df35ded58048d9aa3923774657c@2x.jpg)

In [ ]:
Tempsave["Watch Time"] = Tempsave["datetime"].diff().abs()
Tempsave.head()

,Platform,Person,Date,Link,datetime,date,Time,Day,Watch Time
0,Tiktok,A,2026-03-20 12:33:12,https://www.tiktokv.com/share/video/7616782825...,2026-03-20 12:33:12,2026-03-20,12:33:12,Friday,NaT
1,Tiktok,A,2026-03-20 12:32:59,https://www.tiktokv.com/share/video/7618946244...,2026-03-20 12:32:59,2026-03-20,12:32:59,Friday,0 days 00:00:13
2,Tiktok,A,2026-03-20 12:02:37,https://www.tiktokv.com/share/video/7619059457...,2026-03-20 12:02:37,2026-03-20,12:02:37,Friday,0 days 00:30:22
3,Tiktok,A,2026-03-20 12:02:36,https://www.tiktokv.com/share/video/7501170729...,2026-03-20 12:02:36,2026-03-20,12:02:36,Friday,0 days 00:00:01
4,Tiktok,A,2026-03-20 11:53:03,https://www.tiktokv.com/share/video/7618388640...,2026-03-20 11:53:03,2026-03-20,11:53:03,Friday,0 days 00:09:33


Full disclosure, I got so confused with the detla... I went to Claude and was like "please help" and it gave me the .diff().abs() 

So... we've done most of what I set out to do now time to clean it up

In [ ]:
ttdf2a = Tempsave[["Platform", "Person", "date", "Day", "Time", "Watch Time"]]
ttdf2a.head()

,Platform,Person,date,Day,Time,Watch Time
0,Tiktok,A,2026-03-20,Friday,12:33:12,NaT
1,Tiktok,A,2026-03-20,Friday,12:32:59,0 days 00:00:13
2,Tiktok,A,2026-03-20,Friday,12:02:37,0 days 00:30:22
3,Tiktok,A,2026-03-20,Friday,12:02:36,0 days 00:00:01
4,Tiktok,A,2026-03-20,Friday,11:53:03,0 days 00:09:33


![Image](https://i.imgflip.com/6p8gkl.jpg)

## Okay... So, I think what I need to do to streamline it: 
1. data frame out of the "['Watch History'].iloc[15]" within the OG file tiktok gives me 
2. (Potentially change the name of Date to something so I can use Date later but not confuse myself or the system)
3. Add a column for Date (in the case of what i did "date") with the date in day time 
4. Add a column for day
5. Add a column for time but DON'T convert it to date time 
6. Calcuate watch time not with the time detla 
7. Add a column for Platform
8. Add a column for Person
9. Save that final as a data frame. 

I'm hoping that will eliminate the problems I had and how many data frames I had to save to get to a final. 

Still haven't come up with a naming convention so that's going to be done for the next milestone as I bring in other people's data. I'm beta testing on my own right now. 

#### Next Steps
1. Finish collecting data by Thursday 
2. Start making all the Tiktok Watch Time Tables to have by Milestone 2 
3. If I'm feeling really bold start figuring out how to combine them all. 
4. Webscraping at some point by the other three tables need them.... FUN 
5. Start graphing! 

### Milestone 2 Starts here!

Wow, it hasn't even been a week since Milestone two and oh boy I'm finding MORE things to do. HUZZAH! 
First, found out that on the Instagram app you can download files as .json, where as on the desktop you can't. If I ever meet Mark Zuckerberg I will ask him one question only, "what is wrong with you?" 
ANYWAYS, so I was able to get 3/4 tables done... or so I thought. Are the two tables that show watch time done? NOPE. I need to account for log-in time. I... misunderstood Tiktok's column for log-in history.... I thought it meant every time you type your username and password, *not* every time you open the app. So, shout out Dr. Silva for telling me that data was in there so I have to give another look. 

So... I'm still doing my little beta test on how to finalize these tables before I move on to working with other people's data. I figure if I'm the one digging through my own data I'm okay if it gets messy. I know everything that's in there, and I'm actually pretty boring online. It's all pottery videos, real housewives of salt lake city, and a too high screentime. So I'm okay having other people look at this, lol.  

In [ ]:
with open("Julia Beta recommended_topics.json") as j:
    IGrecommendedtopics = json.load(j)

In [ ]:
with open("Julia Beta videos_watched.json") as j:
    IGactivity = json.load(j)

In [ ]:
### IGrecommendedtopics

{'topics_your_topics': [{'title': '',
   'media_map_data': {},
   'string_map_data': {'Name': {'href': '',
     'value': 'Crafts',
     'timestamp': 0}}},
  {'title': '',
   'media_map_data': {},
   'string_map_data': {'Name': {'href': '',
     'value': 'Pottery & Ceramics',
     'timestamp': 0}}},
  {'title': '',
   'media_map_data': {},
   'string_map_data': {'Name': {'href': '',
     'value': 'Musical Performances',
     'timestamp': 0}}},
  {'title': '',
   'media_map_data': {},
   'string_map_data': {'Name': {'href': '',
     'value': 'Foods',
     'timestamp': 0}}},
  {'title': '',
   'media_map_data': {},
   'string_map_data': {'Name': {'href': '',
     'value': 'TV & Movies Celebrities',
     'timestamp': 0}}},
  {'title': '',
   'media_map_data': {},
   'string_map_data': {'Name': {'href': '',
     'value': 'Video Games',
     'timestamp': 0}}},
  {'title': '',
   'media_map_data': {},
   'string_map_data': {'Name': {'href': '',
     'value': 'Fashion Products',
     'timestam

In [ ]:
type(IGrecommendedtopics)

dict

In [ ]:
IGrecommendedtopics.keys()

dict_keys(['topics_your_topics'])

In [ ]:
topics = [item['string_map_data']['Name']['value'] for item in IGrecommendedtopics['topics_your_topics']]

In [ ]:
IG1A = pd.DataFrame(topics, columns=["Topic"])
IG1A

,Topic
0,Crafts
1,Pottery & Ceramics
2,Musical Performances
3,Foods
4,TV & Movies Celebrities
5,Video Games
6,Fashion Products
7,East Asian Food
8,Alcoholic Beverages
9,Baked Goods


In [ ]:
IG1A_final = IG1A.assign(Platform="Instagram", Person="A")[["Platform", "Person"] + IG1A.columns.tolist()]
IG1A_final.head()

,Platform,Person,Topic
0,Instagram,A,Crafts
1,Instagram,A,Pottery & Ceramics
2,Instagram,A,Musical Performances
3,Instagram,A,Foods
4,Instagram,A,TV & Movies Celebrities


In [ ]:
IGactivity

[{'timestamp': 1774902960,
  'media': [],
  'label_values': [{'label': 'Ad library public URL'},
   {'label': 'URL',
    'value': 'https://www.instagram.com/reel/DWhQVSSAd7P/',
    'href': 'https://www.instagram.com/reel/DWhQVSSAd7P/'},
   {'dict': [{'dict': [{'label': 'URL',
        'value': 'https://stan.store/tarakermiet'},
       {'label': 'Name', 'value': 'Tara Kermiet | Burnout + Career Coach'},
       {'label': 'Username', 'value': 'tarakermiet'}],
      'title': ''}],
    'title': 'Owner'}],
  'fbid': '17841400448200474:18107755963665605:1774902960::DYIImpressionsData'},
 {'timestamp': 1774905887,
  'media': [],
  'label_values': [{'label': 'Ad library public URL'},
   {'label': 'URL',
    'value': 'https://www.instagram.com/reel/DWhKrFzCUEV/',
    'href': 'https://www.instagram.com/reel/DWhKrFzCUEV/'},
   {'dict': [{'dict': [{'label': 'URL', 'value': 'https://www.pi.fyi/welcome'},
       {'label': 'Name',
        'value': 'â\xad\x90ï¸\x8f Perfectly Imperfect â\xad\x90ï¸\x8f'},

In [ ]:
IG2A= pd.DataFrame(IGactivity)

In [ ]:
IG2A.head()

,timestamp,media,label_values,fbid
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...


In [ ]:
IG2A.tail()

,timestamp,media,label_values,fbid
1590,1775043092,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18104102995904540:1775043092...
1591,1775268603,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18078362138529592:1775268603...
1592,1775313945,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18091168172146640:1775313945...
1593,1775315270,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17996132753759009:1775315270...
1594,1775347188,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18401097919195000:1775347188...


In [ ]:
type(IG2A["timestamp"])

pandas.core.series.Series

Okay so I need to pull out the time stamp as date time to do everything. I think it's miliseconds.... joy.

In [ ]:
IG2A["timestamp"]=IG2A["timestamp"].dt.microseconds

AttributeError: Can only use .dt accessor with datetimelike values

In [ ]:
IG2A['date'] = pd.to_datetime(IG2A['timestamp'], unit='ms')

In [ ]:
IG2A

,timestamp,media,label_values,fbid,date
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-21 13:01:42.960
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-21 13:01:45.887
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-21 13:01:48.114
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-21 13:01:51.917
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-21 13:01:52.335
...,...,...,...,...,...
1590,1775043092,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18104102995904540:1775043092...,1970-01-21 13:04:03.092
1591,1775268603,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18078362138529592:1775268603...,1970-01-21 13:07:48.603
1592,1775313945,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18091168172146640:1775313945...,1970-01-21 13:08:33.945
1593,1775315270,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17996132753759009:1775315270...,1970-01-21 13:08:35.270


...okay not milliseconds. 

In [ ]:
IG2A['date'] = pd.to_datetime(IG2A['timestamp'], unit='us')

In [ ]:
IG2A

,timestamp,media,label_values,fbid,date
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:29:34.902960
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:29:34.905887
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:29:34.908114
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:29:34.911917
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:29:34.912335
...,...,...,...,...,...
1590,1775043092,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18104102995904540:1775043092...,1970-01-01 00:29:35.043092
1591,1775268603,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18078362138529592:1775268603...,1970-01-01 00:29:35.268603
1592,1775313945,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18091168172146640:1775313945...,1970-01-01 00:29:35.313945
1593,1775315270,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17996132753759009:1775315270...,1970-01-01 00:29:35.315270


Not... microsecond?

In [ ]:
IG2A['date'] = pd.to_datetime(IG2A['timestamp'], unit='ns')
IG2A

,timestamp,media,label_values,fbid,date
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335
...,...,...,...,...,...
1590,1775043092,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18104102995904540:1775043092...,1970-01-01 00:00:01.775043092
1591,1775268603,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18078362138529592:1775268603...,1970-01-01 00:00:01.775268603
1592,1775313945,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18091168172146640:1775313945...,1970-01-01 00:00:01.775313945
1593,1775315270,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17996132753759009:1775315270...,1970-01-01 00:00:01.775315270


So... not Nano seconds

In [ ]:
IG2A['datetime'] = pd.to_datetime(IG2A['timestamp'], unit='s')
IG2A

,timestamp,media,label_values,fbid,date,datetime
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960,2026-03-30 20:36:00
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887,2026-03-30 21:24:47
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114,2026-03-30 22:01:54
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917,2026-03-30 23:05:17
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335,2026-03-30 23:12:15
...,...,...,...,...,...,...
1590,1775043092,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18104102995904540:1775043092...,1970-01-01 00:00:01.775043092,2026-04-01 11:31:32
1591,1775268603,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18078362138529592:1775268603...,1970-01-01 00:00:01.775268603,2026-04-04 02:10:03
1592,1775313945,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18091168172146640:1775313945...,1970-01-01 00:00:01.775313945,2026-04-04 14:45:45
1593,1775315270,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17996132753759009:1775315270...,1970-01-01 00:00:01.775315270,2026-04-04 15:07:50


IT'S SECONDS!!!!! I should have started with that. Also shout out to pandas userguide, I went right along that list to figure out. Okay now to get this the table to match the one for tiktok

In [ ]:
IG2A["Date"]=IG2A["datetime"].dt.date
IG2A.head()

,timestamp,media,label_values,fbid,date,datetime,Date
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960,2026-03-30 20:36:00,2026-03-30
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887,2026-03-30 21:24:47,2026-03-30
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114,2026-03-30 22:01:54,2026-03-30
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917,2026-03-30 23:05:17,2026-03-30
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335,2026-03-30 23:12:15,2026-03-30


In [ ]:
IG2A["Day"]=IG2A["datetime"].dt.day_name()
IG2A.head()

,timestamp,media,label_values,fbid,date,datetime,Date,Day
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960,2026-03-30 20:36:00,2026-03-30,Monday
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887,2026-03-30 21:24:47,2026-03-30,Monday
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114,2026-03-30 22:01:54,2026-03-30,Monday
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917,2026-03-30 23:05:17,2026-03-30,Monday
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335,2026-03-30 23:12:15,2026-03-30,Monday


In [ ]:
IG2A["Watch Time"] = IG2A["datetime"].diff().abs()
IG2A.head()

,timestamp,media,label_values,fbid,date,datetime,Date,Day,Watch Time
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960,2026-03-30 20:36:00,2026-03-30,Monday,NaT
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887,2026-03-30 21:24:47,2026-03-30,Monday,0 days 00:48:47
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114,2026-03-30 22:01:54,2026-03-30,Monday,0 days 00:37:07
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917,2026-03-30 23:05:17,2026-03-30,Monday,0 days 01:03:23
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335,2026-03-30 23:12:15,2026-03-30,Monday,0 days 00:06:58


In [ ]:
IG2A["Time"]=IG2A["datetime"].dt.time
IG2A.head()

,timestamp,media,label_values,fbid,date,datetime,Date,Day,Watch Time,Time
0,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960,2026-03-30 20:36:00,2026-03-30,Monday,NaT,20:36:00
1,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887,2026-03-30 21:24:47,2026-03-30,Monday,0 days 00:48:47,21:24:47
2,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114,2026-03-30 22:01:54,2026-03-30,Monday,0 days 00:37:07,22:01:54
3,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917,2026-03-30 23:05:17,2026-03-30,Monday,0 days 01:03:23,23:05:17
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335,2026-03-30 23:12:15,2026-03-30,Monday,0 days 00:06:58,23:12:15


In [ ]:
Makeitwork = IG2A.assign(Platform="Instagram", Person="A")[['Platform', 'Person'] + IG2A.columns.tolist()]
Makeitwork.head()

,Platform,Person,timestamp,media,label_values,fbid,date,datetime,Date,Day,Watch Time,Time
0,Instagram,A,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960,2026-03-30 20:36:00,2026-03-30,Monday,NaT,20:36:00
1,Instagram,A,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887,2026-03-30 21:24:47,2026-03-30,Monday,0 days 00:48:47,21:24:47
2,Instagram,A,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114,2026-03-30 22:01:54,2026-03-30,Monday,0 days 00:37:07,22:01:54
3,Instagram,A,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917,2026-03-30 23:05:17,2026-03-30,Monday,0 days 01:03:23,23:05:17
4,Instagram,A,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335,2026-03-30 23:12:15,2026-03-30,Monday,0 days 00:06:58,23:12:15


In [ ]:
IDidit = Makeitwork[["Platform", "Person", "Date", "Day", "Time", "Watch Time"]]
IDidit.head()

,Platform,Person,Date,Day,Time,Watch Time
0,Instagram,A,2026-03-30,Monday,20:36:00,NaT
1,Instagram,A,2026-03-30,Monday,21:24:47,0 days 00:48:47
2,Instagram,A,2026-03-30,Monday,22:01:54,0 days 00:37:07
3,Instagram,A,2026-03-30,Monday,23:05:17,0 days 01:03:23
4,Instagram,A,2026-03-30,Monday,23:12:15,0 days 00:06:58


Julia Notes: 
3 data frames to get to final for Instagram 
I think 3 for Tiktok 
2 for Instagram 

### For Replication: 
*This is a work in progress*
Instagram Table 1: 
Naming convention: 
First, 
Final needs to be IG1(thenletter)

Steps 
1. with open("FILENAME.json") as j:
    IGrecommendedtopics = json.load(j) 
2. topics = [item['string_map_data']['Name']['value'] for item in IGrecommendedtopics['topics_your_topics']]
3. 

Tiktok Table 1: 
Naming convention: 
Final needs to be TT1(thenletter)

Instagram Table 2: 

Tiktok Table 2: 

In [ ]:
## df_tiktok["Login History"].iloc[6]

In [ ]:
TTlogin= pd.DataFrame(df_tiktok["Login History"].iloc[6])

In [ ]:
TTlogin.head()

,Date,IP,DeviceModel,DeviceSystem,NetworkType,Carrier
0,2024-12-26 00:01:54,75.38.153.117,"iPhone17,2",iOS 18.1.1,Wi-Fi,
1,2024-12-26 00:11:53,75.38.153.117,"iPhone17,2",iOS 18.1.1,Wi-Fi,
2,2024-12-26 00:16:17,75.38.153.117,"iPhone17,2",iOS 18.1.1,Wi-Fi,
3,2024-12-26 00:22:33,75.38.153.117,"iPhone17,2",iOS 18.1.1,Wi-Fi,
4,2024-12-26 00:23:45,75.38.153.117,"iPhone17,2",iOS 18.1.1,Wi-Fi,


In [ ]:
type(TTlogin["Date"])

pandas.core.series.Series

It's never datetime. That's my number one take away from IG and TT downloadable data. 

Notes from talking to Dr. Silva in class: OKAY! Changing things up. 
Concat- Post view time and video view time, cron. order. 
Take Log in time 

For tiktok 
df_logins 
df_watches in 

df.watches.iloc [new.index -1 ]

### Below is IG Post View Time

In [ ]:
with open("Julia Beta posts_viewed.json") as j:
    IGpostviewed = json.load(j)
IGpostviewed

[{'timestamp': 1774909773,
  'media': [],
  'label_values': [{'label': 'Ad library public URL'},
   {'label': 'URL',
    'value': 'https://www.instagram.com/p/DWgxvTrlPfF/',
    'href': 'https://www.instagram.com/p/DWgxvTrlPfF/'},
   {'dict': [{'dict': [{'label': 'URL', 'value': 'https://likeshop.me/tmz_tv'},
       {'label': 'Name', 'value': 'TMZ'},
       {'label': 'Username', 'value': 'tmz_tv'}],
      'title': ''}],
    'title': 'Owner'}],
  'fbid': '17841400448200474:18591305350002568:1774909773::DYIImpressionsData'},
 {'timestamp': 1774909773,
  'media': [],
  'label_values': [{'label': 'Ad library public URL'},
   {'label': 'URL',
    'value': 'https://www.instagram.com/p/DWhcKrhifPf/',
    'href': 'https://www.instagram.com/p/DWhcKrhifPf/'},
   {'dict': [{'dict': [{'label': 'URL', 'value': 'http://kalesaladfamous.com'},
       {'label': 'Name', 'value': 'Memes'},
       {'label': 'Username', 'value': 'kalesalad'}],
      'title': ''}],
    'title': 'Owner'}],
  'fbid': '1784140

In [ ]:
PostViewHistory = pd.DataFrame(IGpostviewed)
PostViewHistory.head()

,timestamp,media,label_values,fbid
0,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18591305350002568:1774909773...
1,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18187645741372271:1774909773...
2,1774910577,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18085570352350779:1774910577...
3,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17932913811218141:1774912335...
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18098571358814201:1774912335...


...Well it's good that the layout is the same as the video one

In [ ]:
PostViewHistory['datetime'] = pd.to_datetime(PostViewHistory['timestamp'], unit='s')
PostViewHistory.head()

,timestamp,media,label_values,fbid,datetime
0,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18591305350002568:1774909773...,2026-03-30 22:29:33
1,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18187645741372271:1774909773...,2026-03-30 22:29:33
2,1774910577,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18085570352350779:1774910577...,2026-03-30 22:42:57
3,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17932913811218141:1774912335...,2026-03-30 23:12:15
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18098571358814201:1774912335...,2026-03-30 23:12:15


In [ ]:
PostViewHistory["Time"]=PostViewHistory["datetime"].dt.time
PostViewHistory.head()

,timestamp,media,label_values,fbid,datetime,Time
0,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18591305350002568:1774909773...,2026-03-30 22:29:33,22:29:33
1,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18187645741372271:1774909773...,2026-03-30 22:29:33,22:29:33
2,1774910577,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18085570352350779:1774910577...,2026-03-30 22:42:57,22:42:57
3,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17932913811218141:1774912335...,2026-03-30 23:12:15,23:12:15
4,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18098571358814201:1774912335...,2026-03-30 23:12:15,23:12:15


In [ ]:
PostViewHistorywithLabels = PostViewHistory.assign(Posting="Post", Person="A")[["Posting", "Person"] + PostViewHistory.columns.tolist()]
PostViewHistorywithLabels.head()

,Posting,Person,timestamp,media,label_values,fbid,datetime,Time
0,Post,A,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18591305350002568:1774909773...,2026-03-30 22:29:33,22:29:33
1,Post,A,1774909773,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18187645741372271:1774909773...,2026-03-30 22:29:33,22:29:33
2,Post,A,1774910577,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18085570352350779:1774910577...,2026-03-30 22:42:57,22:42:57
3,Post,A,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17932913811218141:1774912335...,2026-03-30 23:12:15,23:12:15
4,Post,A,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18098571358814201:1774912335...,2026-03-30 23:12:15,23:12:15


In [ ]:
ReadyToConCatIGPost = PostViewHistorywithLabels[["Posting", "Person", "datetime", "Time"]]
ReadyToConCatIGPost.head()

,Posting,Person,datetime,Time
0,Post,A,2026-03-30 22:29:33,22:29:33
1,Post,A,2026-03-30 22:29:33,22:29:33
2,Post,A,2026-03-30 22:42:57,22:42:57
3,Post,A,2026-03-30 23:12:15,23:12:15
4,Post,A,2026-03-30 23:12:15,23:12:15


### Great now to get the table to mirror that for video


In [ ]:
AlmostReadyToConCatIGVideo= Makeitwork.assign(Posting="Video")[["Posting"] + Makeitwork.columns.tolist()]
AlmostReadyToConCatIGVideo.head()

,Posting,Platform,Person,timestamp,media,label_values,fbid,date,datetime,Date,Day,Watch Time,Time
0,Video,Instagram,A,1774902960,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18107755963665605:1774902960...,1970-01-01 00:00:01.774902960,2026-03-30 20:36:00,2026-03-30,Monday,NaT,20:36:00
1,Video,Instagram,A,1774905887,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18310119415256128:1774905887...,1970-01-01 00:00:01.774905887,2026-03-30 21:24:47,2026-03-30,Monday,0 days 00:48:47,21:24:47
2,Video,Instagram,A,1774908114,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18578712475019167:1774908114...,1970-01-01 00:00:01.774908114,2026-03-30 22:01:54,2026-03-30,Monday,0 days 00:37:07,22:01:54
3,Video,Instagram,A,1774911917,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:18062489216380208:1774911917...,1970-01-01 00:00:01.774911917,2026-03-30 23:05:17,2026-03-30,Monday,0 days 01:03:23,23:05:17
4,Video,Instagram,A,1774912335,[],"[{'label': 'Ad library public URL'}, {'label':...",17841400448200474:17952887547103349:1774912335...,1970-01-01 00:00:01.774912335,2026-03-30 23:12:15,2026-03-30,Monday,0 days 00:06:58,23:12:15


In [ ]:
ReadyToConCatIGVideo = AlmostReadyToConCatIGVideo[["Posting","Person", "datetime", "Time"]]
ReadyToConCatIGVideo.head()

,Posting,Person,datetime,Time
0,Video,A,2026-03-30 20:36:00,20:36:00
1,Video,A,2026-03-30 21:24:47,21:24:47
2,Video,A,2026-03-30 22:01:54,22:01:54
3,Video,A,2026-03-30 23:05:17,23:05:17
4,Video,A,2026-03-30 23:12:15,23:12:15


Took out the IG tag and putting in Video. Maybe I need IG but tbh I'm more worried about Concatting lol
Okay my goal is to concat AND put the date time in chronological order. 

In [ ]:
FULLIGWatchHistory = pd.concat([ReadyToConCatIGPost, ReadyToConCatIGVideo]).sort_values("datetime")
FULLIGWatchHistory.head()

,Posting,Person,datetime,Time
5,Video,A,2026-03-24 01:19:10,01:19:10
6,Video,A,2026-03-24 20:45:00,20:45:00
38,Post,A,2026-03-25 01:10:44,01:10:44
39,Post,A,2026-03-25 01:41:36,01:41:36
40,Post,A,2026-03-25 03:03:20,03:03:20


In [ ]:
FULLIGWatchHistory.tail()

,Posting,Person,datetime,Time
1250,Post,A,2026-04-06 19:54:03,19:54:03
1246,Post,A,2026-04-06 19:54:03,19:54:03
1247,Post,A,2026-04-06 19:54:03,19:54:03
1249,Post,A,2026-04-06 19:54:03,19:54:03
1248,Post,A,2026-04-06 19:54:03,19:54:03


DID IT WORK? I THINK SO. I got very confused with the concat index = true portion but I left it out and it looks.. fine? The colum on the left hand side is totally throwing me but OKAY

Now I need to make the log in time a date time. JOY! 
Okay thing to note.... I did not originally download Log in information so this one is going to be wonky. The other ones will NOT be wonky. 

In [ ]:
with open("Julia Beta_login_activity.json") as j:
    IGLoginInfo = json.load(j)
## IGLoginInfo

{'account_history_login_history': [{'title': '2026-03-20T20:29:25+00:00',
   'media_map_data': {},
   'string_map_data': {'Cookie Name': {'href': '',
     'value': '*************************yOy',
     'timestamp': 0},
    'IP Address': {'href': '',
     'value': '2600:1702:3020:20f0:55f1:6175:faa8:1343',
     'timestamp': 0},
    'Port': {'href': '', 'value': '41130', 'timestamp': 0},
    'Language Code': {'href': '', 'value': 'en', 'timestamp': 0},
    'Time': {'href': '', 'value': '', 'timestamp': 1774038565},
    'User Agent': {'href': '',
     'value': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Safari/537.36',
     'timestamp': 0}}},
  {'title': '2026-03-20T20:28:42+00:00',
   'media_map_data': {},
   'string_map_data': {'Cookie Name': {'href': '',
     'value': '*************************iO8',
     'timestamp': 0},
    'IP Address': {'href': '',
     'value': '2600:1702:3020:20f0:8927:6f5d:ecfc:d650',
     'timestamp': 0

In [ ]:
IGLoginInfo.keys()
IGLoginInfo['account_history_login_history'][0]['string_map_data']["Time"]['timestamp']

1774038565

In [ ]:
times = [x['string_map_data']["Time"]['timestamp'] for x in IGLoginInfo['account_history_login_history']]

IGLogInTime= pd.DataFrame({'login_information': pd.to_datetime(times, unit='s')})
IGLogInTime['login_information']

0    2026-03-20 20:29:25
1    2026-03-20 20:28:42
2    2026-03-20 20:28:16
3    2026-03-19 22:56:09
4    2026-03-19 22:52:23
             ...        
63   2017-02-23 19:24:04
64   2017-02-23 19:23:40
65   2017-02-23 01:34:18
66   2017-02-10 05:50:36
67   2016-12-04 20:30:05
Name: login_information, Length: 68, dtype: datetime64[ns]

In [ ]:
IGLogInTime['login_information'].iloc[0]

Timestamp('2026-03-20 20:29:25')

In [ ]:
FULLIGWatchHistory["datetime"] in ['2017-02-23 19:24:04']

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [ ]:
FULLIGWatchHistory[FULLIGWatchHistory["datetime"] in list(IGLogInTime["login_information"])]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

SO I'm stuck........... so I'm going to make a Heat Graph draft to keep the progress moving forward... I just want to figure out how to do it. 

So alot of these heat maps are.... simple. https://stackoverflow.com/questions/71549352/plotly-express-heatmap-using-pandas-dataframe FUN

#### We did in fact NOT do it, Joe. 
But we're getting close.

Next steps: 
1. String-ify date time to use the x for x in to get the actual watch time. Finalize that table. 
2. Copy and Paste all the code that I actually need into a reference document so I can start on a clean slate with student data and have a template to work with to make the most of my time. 
3. A draft of a heat map, the can I am forever kicking down the road but NO LONGER. 